# Introductory H2O Machine Learning Tutorial

Prepared for H2O Open Chicago 2016: http://open.h2o.ai/chicago.html

## Install H2O

The first step in this tutorial is to download and install the h2o Python module.  
The latest version is always here: http://www.h2o.ai/download/h2o/py

### Start up the H2O Cluster

Once the Python module is installed, we begin by starting up a local (on your laptop) H2O cluster.

In [1]:
# Load the H2O library and start up the H2O cluter locally on your machine
import h2o

# Number of threads, nthreads = -1, means use all cores on your machine
# max_mem_size is the maximum memory (in GB) to allocate to H2O
h2o.init(nthreads = -1, max_mem_size = 8)



No instance found at ip and port: localhost:54321. Trying to start local jar...


JVM stdout: /var/folders/2j/jg4sl53d5q53tc2_nzm9fz5h0000gn/T/tmp2QMc58/h2o_me_started_from_python.out
JVM stderr: /var/folders/2j/jg4sl53d5q53tc2_nzm9fz5h0000gn/T/tmpgB7EFj/h2o_me_started_from_python.err
Using ice_root: /var/folders/2j/jg4sl53d5q53tc2_nzm9fz5h0000gn/T/tmp2jVpCf


Java Version: java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)


Starting H2O JVM and connecting: ......... Connection successful!


/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:92: DeprecationWarning: DisplayFormatter._ipython_display_formatter_default is deprecated: use @default decorator instead.
  def _ipython_display_formatter_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:98: DeprecationWarning: DisplayFormatter._formatters_default is deprecated: use @default decorator instead.
  def _formatters_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:677: DeprecationWarning: PlainTextFormatter._deferred_printers_default is deprecated: use @default decorator instead.
  def _deferred_printers_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:669: DeprecationWarning: PlainTextFormatter._singleton_printers_default is deprecated: use @default decorator instead.
  def _singleton_printers_default(self):
/usr/local/lib/python2.7/site-packages/IPython/core/formatters.py:672: DeprecationWarning: PlainTextFormat

------------------------------  ---------------------------------
H2O cluster uptime:             1 seconds 34 milliseconds
H2O cluster version:            3.8.2.3
H2O cluster name:               H2O_started_from_python_me_tva596
H2O cluster total nodes:        1
H2O cluster total free memory:  7.11 GB
H2O cluster total cores:        8
H2O cluster allowed cores:      8
H2O cluster healthy:            True
H2O Connection ip:              127.0.0.1
H2O Connection port:            54321
H2O Connection proxy:
Python Version:                 2.7.10
------------------------------  ---------------------------------

## Data prep

### Import data
Next we will import a cleaned up version of the Lending Club "Bad Loans" dataset. The purpose here is to predict whether a loan will be bad (i.e. not repaid to the lender). The response column, `bad_loan`, is 1 if the loan was bad, and 0 otherwise.

In [2]:
loan_csv = "/Volumes/H2OTOUR/loan.csv"  # modify this for your machine
# Alternatively, you can import the data directly from a URL
#loan_csv = "https://raw.githubusercontent.com/h2oai/app-consumer-loan/master/data/loan.csv"
data = h2o.import_file(loan_csv)  # 163,987 rows x 15 columns


Parse Progress: [                                                  ] 00%

/usr/local/lib/python2.7/site-packages/h2o/h2o.py:316: UserWarning: ParseError at file nfs://Volumes/H2OTOUR/loan.csv  at byte offset 4194304; error = 'Unmatched quote char "'
  warnings.warn(w)


Parse Progress: [##################################################] 100%


In [3]:
data.shape

(163987, 15)

### Encode response variable
Since we want to train a binary classification model, we must ensure that the response is coded as a factor. If the response is 0/1, H2O will assume it's numeric, which means that H2O will train a regression model instead.

In [4]:
data['bad_loan'] = data['bad_loan'].asfactor()  #encode the binary repsonse as a factor
data['bad_loan'].levels()  #optional: after encoding, this shows the two factor levels, '0' and '1'

[['0', '1']]

### Partition data

Next, we partition the data into training, validation and test sets.

In [5]:
# Partition data into 70%, 15%, 15% chunks
# Setting a seed will guarantee reproducibility

splits = data.split_frame(ratios=[0.7, 0.15], seed=1)  

train = splits[0]
valid = splits[1]
test = splits[2]

Notice that `split_frame()` uses approximate splitting not exact splitting (for efficiency), so these are not exactly 70%, 15% and 15% of the total rows.

In [6]:
print train.nrow
print valid.nrow
print test.nrow

114908
24498
24581


### Identify response and predictor variables
In H2O, we use `y` to designate the response variable and `x` to designate the list of predictor columns.

In [7]:
y = 'bad_loan'
x = list(data.columns)

In [8]:
x.remove(y)  #remove the response
x.remove('int_rate')  #remove the interest rate column because it's correlated with the outcome

In [9]:
# List of predictor columns
x

[u'loan_amnt',
 u'term',
 u'emp_length',
 u'home_ownership',
 u'annual_inc',
 u'purpose',
 u'addr_state',
 u'dti',
 u'delinq_2yrs',
 u'revol_util',
 u'total_acc',
 u'longest_credit_length',
 u'verification_status']

## H2O Machine Learning

Now that we have prepared the data, we can train some models. We will start by training a single model from each of the H2O supervised algos:

- Generalized Linear Model (GLM)
- Random Forest (RF)
- Gradient Boosting Machine (RF)
- Deep Learning (DL)
- Naive Bayes (NB)

## 1. Generalized Linear Model
Let's start with a basic binomial Generalized Linear Model (GLM).  By default, H2O's GLM uses a regularized, elastic net model.

In [10]:
# Import H2O GLM:
from h2o.estimators.glm import H2OGeneralizedLinearEstimator

### Train a default GLM
We first create an object of class, `"H2OGeneralizedLinearEstimator"`.  This does not actually do any training, it just sets the model up for training by specifying model parameters.

In [11]:
# Initialize the GLM estimator:
# Similar to R's glm() and H2O's R GLM, H2O's GLM has the "family" argument

glm_fit1 = H2OGeneralizedLinearEstimator(family='binomial', model_id='glm_fit1')

Now that `glm_fit1` object is initialized, we can train the model:

In [12]:
glm_fit1.train(x=x, y=y, training_frame=train)


glm Model Build Progress: [##################################################] 100%


### Train a GLM with lambda search

Next we will do some automatic tuning by passing in a validation frame and setting `lambda_search = True`.  Since we are training a GLM with regularization, we should try to find the right amount of regularization (to avoid overfitting).  The model parameter, `lambda`, controls the amount of regularization in a GLM model and we can find the optimal value for `lambda` automatically by setting `lambda_search = True` and passing in a validation frame (which is used to evaluate model performance using a particular value of lambda).

In [13]:
glm_fit2 = H2OGeneralizedLinearEstimator(family='binomial', model_id='glm_fit2', lambda_search=True)
glm_fit2.train(x=x, y=y, training_frame=train, validation_frame=valid)


glm Model Build Progress: [##################################################] 100%


### Evaluate model performance
Let's compare the performance of the two GLMs that were just trained.

In [14]:
glm_perf1 = glm_fit1.model_performance(test)
glm_perf2 = glm_fit2.model_performance(test)

In [15]:
# Print model performance
print glm_perf1
print glm_perf2


ModelMetricsBinomialGLM: glm
** Reported on test data. **

MSE: 0.14215506685
R^2: 0.0639172702162
LogLoss: 0.451078449301
Null degrees of freedom: 24580
Residual degrees of freedom: 24529
Null deviance: 23672.9222656
Residual deviance: 22175.9187245
AIC: 22279.9187245
AUC: 0.677449084114
Gini: 0.354898168228

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.193355175264: 


0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      13646  6345  0.3174   (6345.0/19991.0)
1      1939   2651  0.4224   (1939.0/4590.0)
Total  15585  8996  0.337    (8284.0/24581.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                      threshold    value     idx
--------------------------  -----------  --------  -----
max f1                      0.193355     0.390255  225
max f2                      0.118771     0.556655  306
max f0point5                0.277629     0.353861  149
max accuracy                0.494239     0.81441   32
max precision               0.744507     1         0
max recall                  0.00257537   1         398
max specificity             0.744507     1         0
max absolute_MCC            0.1984       0.210699  220
max min_per_class_accuracy  0.180095     0.627983  238


Gains/Lift Table: Avg response rate: 18.67 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100077                   0.465373           2.8736    2.8736             0.536585         0.536585                    0.0287582       0.0287582                  187.36    187.36
    2        0.0200155                   0.428698           2.39466   2.63413            0.447154         0.49187                     0.0239651       0.0527233                  139.466   163.413
    3        0.0300232                   0.402118           2.24228   2.50351            0.418699         0.46748                     0.0224401       0.0751634                  124.228   150.351
    4        0.0400309                   0.385445           2.28581   2.44909            0.426829         0.457317                    0.0228758       0.0980392                  128.581   144.909
    5        0.0500386                   0.369289           2.04635   2.36854            0.382114         0.442276                    0.0204793       0.118519                   104.635   136.854
    6        0.100037                    0.309726           1.89114   2.12994            0.353133         0.397723                    0.0945534       0.213072                   89.1144   112.994
    7        0.150035                    0.274385           1.68634   1.98211            0.31489          0.370119                    0.0843137       0.297386                   68.6343   98.2114
    8        0.200033                    0.248213           1.39439   1.83521            0.260374         0.342689                    0.0697168       0.367102                   39.4392   83.5213
    9        0.300028                    0.210783           1.27892   1.64981            0.238812         0.308068                    0.127887        0.494989                   27.8919   64.9807
    10       0.400024                    0.184451           1.15473   1.52605            0.215622         0.284959                    0.115468        0.610458                   15.4731   52.6051
    11       0.50002                     0.162093           0.91507   1.40386            0.170871         0.262143                    0.0915033       0.701961                   -8.49301  40.3864
    12       0.600016                    0.14267            0.825742  1.30752            0.15419          0.244152                    0.0825708       0.784532                   -17.4258  30.7517
    13       0.700012                    0.124325           0.753843  1.22843            0.140765         0.229383                    0.0753813       0.859913                   -24.6157  22.8426
    14       0.800008                    0.10547            0.566472  1.14569            0.105777         0.213933                    0.0566449       0.916558                   -43.3528  14.5686
    15       0.900004                    0.0829515          0.509825  1.07504            0.0951993        0.200741                    0.0509804       0.967538                   -49.0175  7.50375
    16       1                           3.3495e-16         0.324632  1                  0.0606184        0.18673                     0.0324619       1                          -67.5368  0




ModelMetricsBinomialGLM: glm
** Reported on test data. **

MSE: 0.142127171974
R^2: 0.0641009563279
LogLoss: 0.450988500491
Null degrees of freedom: 24580
Residual degrees of freedom: 24517
Null deviance: 23672.9222656
Residual deviance: 22171.4966611
AIC: 22299.4966611
AUC: 0.677675858276
Gini: 0.355351716551

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.192945688329: 


0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      13596  6395  0.3199   (6395.0/19991.0)
1      1928   2662  0.42     (1928.0/4590.0)
Total  15524  9057  0.3386   (8323.0/24581.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                      threshold    value     idx
--------------------------  -----------  --------  -----
max f1                      0.192946     0.390122  220
max f2                      0.118948     0.556655  306
max f0point5                0.274462     0.354242  146
max accuracy                0.496581     0.81441   29
max precision               0.745405     1         0
max recall                  0.00276381   1         398
max specificity             0.745405     1         0
max absolute_MCC            0.198282     0.210585  215
max min_per_class_accuracy  0.180234     0.629847  234


Gains/Lift Table: Avg response rate: 18.67 %



group    cumulative_data_fraction    lower_threshold    lift      cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  --------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100077                   0.465573           2.89537   2.89537            0.54065          0.54065                     0.028976        0.028976                   189.537   189.537
    2        0.0200155                   0.429301           2.39466   2.64501            0.447154         0.493902                    0.0239651       0.0529412                  139.466   164.501
    3        0.0300232                   0.402523           2.26405   2.51802            0.422764         0.47019                     0.022658        0.0755991                  126.405   151.802
    4        0.0400309                   0.385086           2.26405   2.45453            0.422764         0.458333                    0.022658        0.0982571                  126.405   145.453
    5        0.0500386                   0.369743           1.95927   2.35548            0.365854         0.439837                    0.0196078       0.117865                   95.927    135.548
    6        0.100037                    0.309986           1.90857   2.13212            0.356387         0.398129                    0.0954248       0.21329                    90.8574   113.212
    7        0.150035                    0.274591           1.69942   1.98792            0.317331         0.371204                    0.0849673       0.298257                   69.9416   98.7922
    8        0.200033                    0.248333           1.41182   1.84393            0.263629         0.344316                    0.0705882       0.368845                   41.1822   84.3927
    9        0.300028                    0.21088            1.27456   1.65416            0.237998         0.308881                    0.127451        0.496296                   27.4562   65.4164
    10       0.400024                    0.184542           1.15255   1.52877            0.215216         0.285467                    0.115251        0.611547                   15.2552   52.8774
    11       0.50002                     0.162026           0.917249  1.40648            0.171277         0.262631                    0.0917211       0.703268                   -8.27514  40.6479
    12       0.600016                    0.14272            0.830099  1.31042            0.155004         0.244695                    0.0830065       0.786275                   -16.9901  31.0422
    13       0.700012                    0.124235           0.738592  1.22874            0.137917         0.229442                    0.0738562       0.860131                   -26.1408  22.8737
    14       0.800008                    0.105495           0.570829  1.1465             0.106591         0.214086                    0.0570806       0.917211                   -42.9171  14.6503
    15       0.900004                    0.0829093          0.503288  1.07504            0.0939788        0.200741                    0.0503268       0.967538                   -49.6712  7.50375
    16       1                           3.25106e-16        0.324632  1                  0.0606184        0.18673                     0.0324619       1                          -67.5368  0

Instead of printing the entire model performance metrics object, it is probably easier to print just the metric that you are interested in comparing.

In [16]:
# Retreive test set AUC
print glm_perf1.auc()
print glm_perf2.auc()

0.677449084114
0.677675858276


In [17]:
# Compare test AUC to the training AUC and validation AUC
print glm_fit2.auc(train=True)
print glm_fit2.auc(valid=True)

0.674306164325
0.675512216705


## 2. Random Forest
H2O's Random Forest (RF) is implements a distributed version of the standard Random Forest algorithm and variable importance measures.

In [18]:
# Import H2O RF:
from h2o.estimators.random_forest import H2ORandomForestEstimator

### Train a default RF
First we will train a basic Random Forest model with default parameters. Random Forest will infer the response distribution from the response encoding. A seed is required for reproducibility.

In [19]:
# Initialize the RF estimator:

rf_fit1 = H2ORandomForestEstimator(model_id='rf_fit1', seed=1)

Now that `rf_fit1` object is initialized, we can train the model:

In [20]:
rf_fit1.train(x=x, y=y, training_frame=train)


drf Model Build Progress: [##################################################] 100%


### Train an RF with more trees

Next we will increase the number of trees used in the forest by setting `ntrees = 100`.  The default number of trees in an H2O Random Forest is 50, so this RF will be twice as big as the default.  Usually increasing the number of trees in an RF will increase performance as well.  Unlike Gradient Boosting Machines (GBMs), Random Forests are fairly resistant (although not free from) overfitting by increasing the number of trees.  See the GBM example below for additional guidance on preventing overfitting using H2O's early stopping functionality.

In [21]:
rf_fit2 = H2ORandomForestEstimator(model_id='rf_fit2', ntrees=100, seed=1)
rf_fit2.train(x=x, y=y, training_frame=train)


drf Model Build Progress: [##################################################] 100%


### Compare model performance
Let's compare the performance of the two RFs that were just trained.

In [22]:
rf_perf1 = rf_fit1.model_performance(test)
rf_perf2 = rf_fit2.model_performance(test)

In [23]:
# Retreive test set AUC
print rf_perf1.auc()
print rf_perf2.auc()

0.662266990734
0.66525468051


### Cross-validate performance

Rather than using held-out test set to evaluate model performance, a user may wish to estimate model performance using cross-validation.  Using the RF algorithm (with default model parameters) as an example, we demonstrate how to perform k-fold cross-validation using H2O.  No custom code or loops are required, you simply specify the number of desired folds in the `nfolds` argument.

Since we are not going to use a test set here, we can use the original (full) dataset, which we called `data` rather than the subsampled `train` dataset.  Note that this will take approximately k (`nfolds`) times longer than training a single RF model, since it will train k models in the cross-validation process (trained on n(k-1)/k rows), in addition to the final model trained on the full `training_frame` dataset with n rows. 

In [24]:
rf_fit3 = H2ORandomForestEstimator(model_id='rf_fit3', seed=1, nfolds=5)
rf_fit3.train(x=x, y=y, training_frame=data)


drf Model Build Progress: [##################################################] 100%


To evaluate the cross-validated AUC, do the following:

In [25]:
print rf_fit3.auc(xval=True)

0.661201482614


Note that the cross-validated AUC is slighly higher than the test set performance we estimated for `rf_fit1`, and this is likely due to the fact that we trained on more data (n rows) than we did while using `train` as the training set (0.75*n rows) in `rf_fit1`.

## 3. Gradient Boosting Machine
H2O's Gradient Boosting Machine (GBM) offers a Stochastic GBM, which can increase performance quite a bit compared to the original GBM implementation.

In [26]:
# Import H2O GBM:
from h2o.estimators.gbm import H2OGradientBoostingEstimator

### Train a default GBM

First we will train a basic GBM model with default parameters. GBM will infer the response distribution from the response encoding if not specified explicitly through the `distribution` argument. A seed is required for reproducibility.

In [27]:
# Initialize and train the GBM estimator:

gbm_fit1 = H2OGradientBoostingEstimator(model_id='gbm_fit1', seed=1)
gbm_fit1.train(x=x, y=y, training_frame=train)


gbm Model Build Progress: [##################################################] 100%


### Train a GBM with more trees

Next we will increase the number of trees used in the GBM by setting `ntrees=500`.  The default number of trees in an H2O GBM is 50, so this GBM will trained using ten times the default.  Increasing the number of trees in a GBM is one way to increase performance of the model, however, you have to be careful not to overfit your model to the training data by using too many trees.  To automatically find the optimal number of trees, you must use H2O's early stopping functionality.  This example will not do that, however, the following example will.

In [28]:
gbm_fit2 = H2OGradientBoostingEstimator(model_id='gbm_fit2', ntrees=500, seed=1)
gbm_fit2.train(x=x, y=y, training_frame=train)


gbm Model Build Progress: [##################################################] 100%


### Train a GBM with early stopping

We will again set `ntrees = 500`, however, this time we will use early stopping in order to prevent overfitting (from too many trees).  All of H2O's algorithms have early stopping available, however, with the exception of Deep Learning, it is not enabled by default.  

There are several parameters that should be used to control early stopping.  The three that are generic to all the algorithms are: `stopping_rounds`, `stopping_metric` and `stopping_tolerance`.  The stopping metric is the metric by which you'd like to measure performance, and so we will choose AUC here.  The `score_tree_interval` is a parameter specific to Random Forest and GBM.  Setting `score_tree_interval=5` will score the model after every five trees.  The parameters we have set below specify that the model will stop training after there have been three scoring intervals where the AUC has not increased more than 0.0005.  Since we have specified a validation frame, the stopping tolerance will be computed on validation AUC rather than training AUC. 

In [29]:
# Now let's use early stopping to find optimal ntrees

gbm_fit3 = H2OGradientBoostingEstimator(model_id='gbm_fit3', 
                                        ntrees=500, 
                                        score_tree_interval=5,     #used for early stopping
                                        stopping_rounds=3,         #used for early stopping
                                        stopping_metric='AUC',     #used for early stopping
                                        stopping_tolerance=0.0005, #used for early stopping
                                        seed=1)

# The use of a validation_frame is recommended with using early stopping
gbm_fit3.train(x=x, y=y, training_frame=train, validation_frame=valid)


gbm Model Build Progress: [##################################################] 100%


### Compare model performance

Let's compare the performance of the three GBMs that were just trained.

In [30]:
gbm_perf1 = gbm_fit1.model_performance(test)
gbm_perf2 = gbm_fit2.model_performance(test)
gbm_perf3 = gbm_fit3.model_performance(test)

In [31]:
# Retreive test set AUC
print gbm_perf1.auc()
print gbm_perf2.auc()
print gbm_perf3.auc()

0.682765594191
0.671854616713
0.68309902855


### Scoring History

To examine the scoring history, use the `scoring_history` method on a trained model.  If `score_tree_interval` is not specified, it will score at various intervals, as we can see for `gbm_fit2.scoring_history()` below.  However, regular 5-tree intervals are used for `gbm_fit3.scoring_history()`.  

The `gbm_fit2` was trained only using a training set (no validation set), so the scoring history is calculated for training set performance metrics only.

In [47]:
gbm_fit2.scoring_history()

timestamp    duration  number_of_trees  training_MSE  \
0     2016-05-03 10:49:35   0.001 sec                0      0.148714   
1     2016-05-03 10:49:35   0.079 sec                1      0.147252   
2     2016-05-03 10:49:35   0.142 sec                2      0.146021   
3     2016-05-03 10:49:36   0.208 sec                3      0.144996   
4     2016-05-03 10:49:36   0.288 sec                4      0.144123   
5     2016-05-03 10:49:36   0.390 sec                5      0.143373   
6     2016-05-03 10:49:36   0.547 sec                6      0.142734   
7     2016-05-03 10:49:36   0.731 sec                7      0.142280   
8     2016-05-03 10:49:36   0.961 sec                8      0.141908   
9     2016-05-03 10:49:37   1.195 sec                9      0.141588   
10    2016-05-03 10:49:37   1.440 sec               10      0.141283   
11    2016-05-03 10:49:37   1.688 sec               11      0.141011   
12    2016-05-03 10:49:37   1.946 sec               12      0.140687   
13    2016-05-03 10:49:38   2.214 sec               13      0.140469   
14    2016-05-03 10:49:38   2.497 sec               14      0.140200   
15    2016-05-03 10:49:38   2.792 sec               15      0.139958   
16    2016-05-03 10:49:39   3.180 sec               16      0.139641   
17    2016-05-03 10:49:39   3.507 sec               17      0.139408   
18    2016-05-03 10:49:39   3.822 sec               18      0.139165   
19    2016-05-03 10:49:43   7.836 sec              114      0.131661   
20    2016-05-03 10:49:53  17.468 sec              324      0.125218   
21    2016-05-03 10:50:02  26.971 sec              500      0.121869   

    training_logloss  training_AUC  training_lift  \
0           0.474030      0.500000       1.000000   
1           0.469209      0.657210       2.680144   
2           0.465266      0.664849       2.663689   
3           0.462036      0.667746       2.860431   
4           0.459326      0.669405       2.900994   
5           0.456995      0.672514       3.013793   
6           0.455020      0.674697       3.080660   
7           0.453580      0.675089       3.088117   
8           0.452386      0.675682       3.056254   
9           0.451341      0.676213       3.064215   
10          0.450355      0.677305       3.071435   
11          0.449475      0.678649       3.090803   
12          0.448489      0.680646       3.129824   
13          0.447748      0.681448       3.148167   
14          0.446905      0.683059       3.199288   
15          0.446166      0.684262       3.202491   
16          0.445185      0.686405       3.178202   
17          0.444452      0.687624       3.181708   
18          0.443691      0.689043       3.219985   
19          0.420788      0.734584       4.119475   
20          0.402242      0.770971       4.827585   
21          0.392608      0.788716       5.042888   

    training_classification_error  
0                        0.818255  
1                        0.400599  
2                        0.346703  
3                        0.339689  
4                        0.361158  
5                        0.338366  
6                        0.352987  
7                        0.336130  
8                        0.341943  
9                        0.330917  
10                       0.326792  
11                       0.338654  
12                       0.357338  
13                       0.326879  
14                       0.339080  
15                       0.342204  
16                       0.346608  
17                       0.337775  
18                       0.343675  
19                       0.257040  
20                       0.244674  
21                       0.212144

When early stopping is used, we see that training stopped at 105 trees instead of the full 500.  Since we used a validation set in `gbm_fit3`, both training and validation performance metrics are stored in the scoring history object.  Take a look at the validation AUC to observe that the correct stopping tolerance was enforced.

In [48]:
gbm_fit3.scoring_history()

timestamp    duration  number_of_trees  training_MSE  \
0     2016-05-03 10:50:06   0.001 sec                0      0.148714   
1     2016-05-03 10:50:07   0.252 sec                5      0.143373   
2     2016-05-03 10:50:07   0.588 sec               10      0.141283   
3     2016-05-03 10:50:08   1.056 sec               15      0.139958   
4     2016-05-03 10:50:08   1.584 sec               20      0.138783   
5     2016-05-03 10:50:09   2.179 sec               25      0.137810   
6     2016-05-03 10:50:09   2.820 sec               30      0.137079   
7     2016-05-03 10:50:10   3.522 sec               35      0.136501   
8     2016-05-03 10:50:11   4.292 sec               40      0.135956   
9     2016-05-03 10:50:12   5.093 sec               45      0.135520   
10    2016-05-03 10:50:12   5.976 sec               50      0.135099   
11    2016-05-03 10:50:13   6.864 sec               55      0.134711   
12    2016-05-03 10:50:16   9.652 sec               60      0.134390   
13    2016-05-03 10:50:17  10.690 sec               65      0.134119   
14    2016-05-03 10:50:18  11.920 sec               70      0.133836   
15    2016-05-03 10:50:20  13.202 sec               75      0.133552   
16    2016-05-03 10:50:21  14.556 sec               80      0.133248   
17    2016-05-03 10:50:22  15.947 sec               85      0.133043   

    training_logloss  training_AUC  training_lift  \
0           0.474030      0.500000       1.000000   
1           0.456995      0.672514       3.013793   
2           0.450355      0.677305       3.071435   
3           0.446166      0.684262       3.202491   
4           0.442520      0.691469       3.330029   
5           0.439517      0.697321       3.420935   
6           0.437235      0.701716       3.507056   
7           0.435430      0.705013       3.502272   
8           0.433766      0.708290       3.583609   
9           0.432420      0.710795       3.588393   
10          0.431123      0.713464       3.636238   
11          0.429942      0.715725       3.684084   
12          0.428975      0.717671       3.746282   
13          0.428143      0.719394       3.774989   
14          0.427330      0.720883       3.789343   
15          0.426443      0.722796       3.851542   
16          0.425563      0.724677       3.894603   
17          0.424940      0.725952       3.937663   

    training_classification_error  validation_MSE  validation_logloss  \
0                        0.818255        0.151042            0.479533   
1                        0.338366        0.146227            0.464094   
2                        0.326792        0.144531            0.458612   
3                        0.342204        0.143614            0.455618   
4                        0.347696        0.142869            0.453261   
5                        0.301302        0.142248            0.451328   
6                        0.304139        0.141892            0.450176   
7                        0.313198        0.141634            0.449319   
8                        0.309265        0.141474            0.448720   
9                        0.301250        0.141333            0.448253   
10                       0.307202        0.141164            0.447708   
11                       0.298717        0.141100            0.447464   
12                       0.298656        0.141044            0.447256   
13                       0.296515        0.141015            0.447140   
14                       0.304087        0.140991            0.447037   
15                       0.306384        0.140982            0.446976   
16                       0.294349        0.140996            0.447003   
17                       0.298622        0.140997            0.446998   

    validation_AUC  validation_lift  validation_classification_error  
0         0.500000         1.000000                         0.814597  
1         0.661129         2.618090                         0.365377  
2         0.663678         2.729585   

## 4. Deep Learning

H2O's Deep Learning algorithm is a multilayer feed-forward artificial neural network.  It can also be used to train an autoencoder, however, in the example below we will train a standard supervised prediction model.

In [49]:
# Import H2O DL:
from h2o.estimators.deeplearning import H2ODeepLearningEstimator

### Train a default DL

First we will train a basic DL model with default parameters. DL will infer the response distribution from the response encoding if not specified explicitly through the `distribution` argument.  H2O's DL will not be reproducbible if run on more than a single core, so in this example, the performance metrics below may vary slightly from what you see on your machine.

In H2O's DL, early stopping is enabled by default, so below, it will use the training set and default stopping parameters to perform early stopping.

In [51]:
# Initialize and train the DL estimator:

dl_fit1 = H2ODeepLearningEstimator(model_id='dl_fit1', seed=1)
dl_fit1.train(x=x, y=y, training_frame=train)

------------------------------  --------------------------------------
H2O cluster uptime:             16 minutes 43 seconds 380 milliseconds
H2O cluster version:            3.8.2.3
H2O cluster name:               H2O_started_from_python_me_wzy124
H2O cluster total nodes:        1
H2O cluster total free memory:  7.03 GB
H2O cluster total cores:        8
H2O cluster allowed cores:      8
H2O cluster healthy:            True
H2O Connection ip:              127.0.0.1
H2O Connection port:            54321
H2O Connection proxy:
Python Version:                 2.7.10
------------------------------  --------------------------------------


deeplearning Model Build Progress: [##################################################] 100%


### Train a DL with new architecture and more epochs

Next we will increase the number of epochs used in the GBM by setting `epochs=20` (the default is 10).  Increasing the number of epochs in a deep neural net may increase performance of the model, however, you have to be careful not to overfit your model.  To automatically find the optimal number of epochs, you must use H2O's early stopping functionality.  Unlike the rest of the H2O algorithms, H2O's DL will use early by default, so we will first turn it off in the next example by setting `stopping_rounds=0`, for comparison.

In [36]:
dl_fit2 = H2ODeepLearningEstimator(model_id='dl_fit2', 
                                   epochs=20, 
                                   hidden=[10,10], 
                                   stopping_rounds=0,  #disable early stopping
                                   seed=1)
dl_fit2.train(x=x, y=y, training_frame=train)


deeplearning Model Build Progress: [##################################################] 100%


### Train a DL with early stopping

This example will use the same model parameters as `dl_fit2`, however, we will turn on early stopping and specify the stopping criterion.  We will also pass a validation set, as is recommended for early stopping.

In [37]:
dl_fit3 = H2ODeepLearningEstimator(model_id='dl_fit3', 
                                   epochs=20, 
                                   hidden=[10,10],
                                   score_interval=1,          #used for early stopping
                                   stopping_rounds=3,         #used for early stopping
                                   stopping_metric='AUC',     #used for early stopping
                                   stopping_tolerance=0.0005, #used for early stopping
                                   seed=1)
dl_fit3.train(x=x, y=y, training_frame=train, validation_frame=valid)


deeplearning Model Build Progress: [##################################################] 100%


### Compare model performance

Again, we will compare the model performance of the three models using a test set and AUC.

In [38]:
dl_perf1 = dl_fit1.model_performance(test)
dl_perf2 = dl_fit2.model_performance(test)
dl_perf3 = dl_fit3.model_performance(test)

In [39]:
# Retreive test set AUC
print dl_perf1.auc()
print dl_perf2.auc()
print dl_perf3.auc()

0.679670061767
0.675026229123
0.681501550425


In [40]:
dl_fit3.scoring_history()

timestamp    duration   training_speed     epochs  iterations  \
0    2016-05-03 10:51:34   0.000 sec             None   0.000000           0   
1    2016-05-03 10:51:35   0.288 sec  518093 rows/sec   0.870192           1   
2    2016-05-03 10:51:36   1.315 sec  607058 rows/sec   6.091290           7   
3    2016-05-03 10:51:37   2.338 sec  662289 rows/sec  12.184347          14   
4    2016-05-03 10:51:38   3.433 sec  698969 rows/sec  19.148841          22   
5    2016-05-03 10:51:38   3.638 sec  700242 rows/sec  20.018589          23   
6    2016-05-03 10:51:38   3.710 sec  700029 rows/sec  20.018589          23   

   samples  training_MSE  training_r2  training_logloss  training_AUC  \
0        0           NaN          NaN               NaN           NaN   
1    99992      0.143602     0.049866          0.457001      0.665048   
2   699938      0.143460     0.050810          0.456577      0.667244   
3  1400079      0.150780     0.002375          0.486599      0.674257   
4  2200355      0.144177     0.046065          0.455084      0.676124   
5  2300296      0.143916     0.047794          0.460532      0.676903   
6  2300296      0.150780     0.002375          0.486599      0.674257   

   training_lift  training_classification_error  validation_MSE  \
0            NaN                            NaN             NaN   
1       2.558203                       0.351841        0.143185   
2       2.721493                       0.363875        0.142990   
3       2.667063                       0.408879        0.150563   
4       2.939212                       0.344357        0.144628   
5       3.048072                       0.375607        0.144099   
6       2.667063                       0.408879        0.150563   

   validation_r2  validation_logloss  validation_AUC  validation_lift  \
0            NaN                 NaN             NaN              NaN   
1       0.051933            0.455546        0.668156         2.663807   
2       0.053225            0.455240        0.671613         2.663807   
3       0.003085            0.486139        0.677053         2.311568   
4       0.042382            0.456287        0.676181         2.641792   
5       0.045880            0.461880        0.675448         2.729852   
6       0.003085            0.486139        0.677053         2.311568   

   validation_classification_error  
0                              NaN  
1                         0.345743  
2                         0.358356  
3                         0.363785  
4                         0.357376  
5                         0.340926  
6                         0.363785

## 4. Naive Bayes

The Naive Bayes (NB) algorithm does not usually beat an algorithm like a Random Forest or GBM, however it is still a popular algorithm, especially in the text domain (when your input is text encoded as "Bag of Words", for example).  The Naive Bayes algorithm is for binary or multiclass classification problems only, not regression.  Therefore, your response must be a factor instead of numeric. 

In [41]:
# Import H2O NB:
from h2o.estimators.naive_bayes import H2ONaiveBayesEstimator

### Train a default NB

First we will train a basic NB model with default parameters. 

In [42]:
# Initialize and train the NB estimator:

nb_fit1 = H2ONaiveBayesEstimator(model_id='nb_fit1')
nb_fit1.train(x=x, y=y, training_frame=train)


naivebayes Model Build Progress: [##################################################] 100%


### Train a NB model with Laplace Smoothing

One of the few tunable model parameters for the Naive Bayes algorithm is the amount of Laplace smoothing.  The H2O Naive Bayes model will not use any Laplace smoothing by default.

In [43]:
nb_fit2 = H2ONaiveBayesEstimator(model_id='nb_fit2', laplace=6)
nb_fit2.train(x=x, y=y, training_frame=train)


naivebayes Model Build Progress: [##################################################] 100%


### Compare model performance

We will compare the model performance of the two NB models using test set AUC.

In [44]:
nb_perf1 = nb_fit1.model_performance(test)
nb_perf2 = nb_fit2.model_performance(test)

In [45]:
# Retreive test set AUC
print nb_perf1.auc()
print nb_perf2.auc()

0.651356759779
0.651421140603


In [46]:
h2o.shutdown(prompt=False)